WIP: ENH: Add complex normal random generator

bashtage opened the pull request:
Add complex normal random generator
Test against the existing normal generator
What's new note

closes #9098
xref #6790

bashtage committed:
ENH: Add complex normal random generator

Add complex normal random generator
Test against the existing normal generator
What's new note

closes #9098
xref #6790

charris commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -551,6 +552,17 @@ cdef object discd_array(rk_state *state, rk_discd func, object size, ndarray oa,
     return array
 
+cdef inline void compute_complex(double *rv_r, double *rv_i, double loc_r,
+                                 double loc_i, double var_r, double var_i, double rho) nogil:
```
Line < 80 characters.

charris commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -551,6 +552,17 @@ cdef object discd_array(rk_state *state, rk_discd func, object size, ndarray oa,
                 PyArray_MultiIter_NEXTi(multi, 1)
     return array
 
+cdef inline void compute_complex(double *rv_r, double *rv_i, double loc_r,
```
`compute_complex` is a very generic name that doesn't say anything about the function.

bashtage replied:
Well, there _are_ two hard problems in computer science....

bashtage committed:
CLN: Rename function and enforce 80 char

Rename compute_complex to bivariate_normal
Ensure all lines are wrapped before 80 characters

charris commented on numpy/random/mtrand/numpy.pxd:
```diff
@@ -29,6 +29,29 @@ cdef extern from "numpy/arrayobject.h":
         NPY_NTYPES
         NPY_NOTYPE
 
+        NPY_INT8
```
Why all of these? Many are not defined by numpy.

bashtage commented:
These come from Cython,

https://github.com/cython/cython/blob/master/Cython/Includes/numpy/__init__.pxd#L60



On Tue, Sep 5, 2017 at 7:03 PM Charles Harris <notifications@github.com>
wrote:

> *@charris* commented on this pull request.
> ------------------------------
>
> In numpy/random/mtrand/numpy.pxd
> <https://github.com/numpy/numpy/pull/9561#discussion_r137070821>:
>
> > @@ -29,6 +29,29 @@ cdef extern from "numpy/arrayobject.h":
>          NPY_NTYPES
>          NPY_NOTYPE
>
> +        NPY_INT8
>
> Why all of these? Many are not defined by numpy.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/9561#pullrequestreview-60691135>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFU5Rc_iiFsXHEiJ2FPmlNUpDnM9kLD8ks5sfYz0gaJpZM4O3TxH>
> .
>

bashtage committed:
REF: Remove Cython-esque enums

Remove Cython enums

charris commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -551,6 +552,19 @@ cdef object discd_array(rk_state *state, rk_discd func, object size, ndarray oa,
                 PyArray_MultiIter_NEXTi(multi, 1)
     return array
 
+cdef inline void bivariate_normal(double *rv_r, double *rv_i, double loc_r,
```
Why inline?

bashtage replied:
I used inline here since I saw some timing evidence that it mattered on Windows.

charris commented on numpy/random/tests/test_random.py:
```diff
@@ -929,6 +929,52 @@ def test_zipf(self):
                             [3, 13]])
         assert_array_equal(actual, desired)
 
+    def test_complex_normal(self):
```
This test seems rather restricted. I like hashing rather long sequences to check for repeatability.

charris commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4527,6 +4541,179 @@ cdef class RandomState:
         x.shape = tuple(final_shape)
         return x
 
+    def complex_normal(self, loc=0.0, gamma=1.0, relation=0.0, size=None):
```
This seems unnecessarily restricted, the wikipedia entry uses a multivariate version where `loc` is a vector and `gamma` and `relation` are square arrays. In fact,  multivariate_complex reduces to multivariate normal with the correct definition of covariance. Even with this restricted form, why not call `multivariate_normal`. You could add broadcasting to the latter if that is what you need, although some reshaping would be needed to transform the multivariate output to complex.

Numpy has a complex type and using that would simplify `bivariate_normal`

charris replied:
It is possible the arange the row and columns of the 2d covariance coming from `gamma` and `relation` so that it remains symmetric and positive semidefinite and the result can be viewed as complex.
```
In [11]: b = zeros((4,4))

In [12]: b[::2, ::2] = ones((2,2))

In [13]: b[::2, 1::2] = ones((2,2)) + 1

In [14]: b[1::2, ::2] = ones((2,2)) + 2

In [15]: b[1::2, 1::2] = ones((2,2)) + 3

In [16]: b
Out[16]: 
array([[ 1.,  2.,  1.,  2.],
       [ 3.,  4.,  3.,  4.],
       [ 1.,  2.,  1.,  2.],
       [ 3.,  4.,  3.,  4.]])
```
That is just a change of basis `p,T @ cov @ p`, where p is a permutation.

bashtage replied:
I felt a full `multivariate_complex_normal` would be limiting in the sense that most expressed demand was for scalar complex RVs and going through a full MV implementation would be very slow since it would require a SVD unless there was a fast path similar to what is here now.  I used the idea of moving from `normal`->`complex_normal` as a natural step.

charris commented:
This is a new function and for those there should be a discussion on the mailing list. You should post there with a link to this PR.

charris commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4533,6 +4547,179 @@ cdef class RandomState:
+        ----------
+        loc : complex or array_like of complex
+            Mean of the distribution.
+        gamma : float, complex or array_like of float or complex
```
`gamma` is Hermitian, hence real in the 1d case. It is also required to be >=0 as it is >= `|relation|`.

charris commented:
I think this implementation over complicated. The covariance matrix is easily factored explicitly and can be done in such a way that gamma and relation broadcast to each other. So you then end up with a stack of matrices that can be multiplied times a stack of normals. That can be done with einsum or, with a reshape, np.matmul. The later isn't as fast as it could be yet for broadcast matrix arguments, but in the common case non-broadcast case it will use cblas. For big sample sets that will be a significant speedup, for small ones other factors are likely to dominate.

Is the broadcasting really needed? if not, the cblas route is the way to go.

bashtage commented:
I agree that is it fairly long.  It is virtually identical to any other rv generator since it has three paths

* scalar in, scalar out
* scalar inputs, array out
* array in, array out with broadcasting

For most of generators the latter two call functions and so the code within the generator is relatively simple and only the scalar-scalar as well as the determination of the inputs types is handled in the function. 

The main filling functions are real only and so the code in each of these segments is a single purpose version of these. 

As for broadcasting -- getting rid of this would greatly simplify the generator, but it seemed out of place for a scalar generator to not support broadcasting given the remainder of the API.

charris commented:
I also note that the complex multivariate normal distribution is just another way to parameterize 2nx2n covariance matrices, there is a one-to-one mapping between the two. Given that, I'd rather just call the usual multivariate normal. If that is too slow, that is where we should do the optimization work, perhaps by adding more methods to the computation, say Cholesky decomposition or a straight forward eigh, and maybe some fast paths. I think the svd method was a mistake, but we keep it for backwards compatibility. If we go the eigh direction, adding an rcond argument with the now customary default might also be useful.

pec27 commented:
+1 on using multivariate_normal.

It would seem you could avoid increasing the `mtrand.pyx` footprint and avoid some bit rot if you used `multivariate_normal(...)` and then convert to complex with a `np.view(...)`.

bashtage commented:
I posted to the mailing list.  For those who prefer github (a shortened version):

There are three paths I could see for this RV:

1. The implementation in the PR following the standard template for scalar RV generators with three paths: scalar->scalar, scalar->array and array->array.  
 - Bulky but identical API to all other scalar double preocsions RV

2. Use `multivariate_normal` and only allow the scalar->scalar and the scalar->array paths.  Possibly could be extended to allow broadcasting if multivariate_normal was extended to allow broadcasting, although I'm not sure of a 100% reliable way reshape broadcastable scalars to have the requirements that a broadcastable `multivairate_normal` would have (covariance being `(2,2,m,n,...)`).

3. It might make more sense to skip a scalar complex normal and just move directly to a
`multivariate_complex_normal`  -- just a higher dimension `multivariate normal`. Does not directly support  a scalar->scalar path, but could be added.

I posted some benchmarks for `normal` vs `multivariate_normal` in the mailing list that show performance close for large array (~10% diff) and very different for scalars (`normal` is about ~30x faster than `multivariate_normal` for scalar generation).

bashtage committed:
REF: Add initial mv normal based gen

Add a multivariate normal-based complex generator for testing and
benchmarking purposes

[skip ci]

charris commented:
Another thought I had last night was to add a `factored` keyword to `multivariate_normal` so that a factored covariance could be passed in, avoiding the factorization step on each call. Maybe add a helper function that can do the factorization using various methods. Or maybe have a lower level routine that simply takes takes a factor, or broadcastable factors.

bashtage commented:
This is now a WIP.  I added a basic complex that uses the multivariate generator.  It is pretty simple, which is good.  The perf is not very good.

```python
In [1]: %timeit np.random.complex_normal(size=1)
3.89 µs ± 13.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [2]: %timeit np.random.complex_normal2(size=1)
135 µs ± 391 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [3]: %timeit np.random.complex_normal(size=1000000)
127 ms ± 727 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [4]: %timeit np.random.complex_normal2(size=1000000)
147 ms ± 900 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

For large arrays, it is about 15% slower.  For scalars it is 30x slower.

The the overhead appear to be about 10µs, and o most of the difference entire difference is down to `multivariate_normal`:

```python
In [11]: %timeit np.random.multivariate_normal(loc,cov)
123 µs ± 881 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [12]: %timeit np.random.multivariate_normal(loc,cov,size=1000000)
150 ms ± 687 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

pec27 commented:
Maybe I am biased by my use-cases, but 15% slowdown for the large simulation doesn’t sound too bad to me. 

If you are going to add a `factored` keyword I would probably call it `cholesky` instead, as that is probably more intuitive.

bashtage commented:
`cholesky` might be misleading since there are other choices for factorization.

pec27 commented:
True enough, but factored sounds a bit more ambiguous to my ears (i.e. the user may not immediately assume what is being factored), whereas if you write cholesky in a context of multivariate normals then it is pretty intuitive (to me at least) that it is going to be the covariance that you are giving the cholesky decomposition of. Also you can probably take advantage of the triangular property, thought that might be early optimization at this point. It was just a suggestion, your mileage may vary.

charris commented:
I intended that factored mean factored, i.e., the covariance has been factored as LL^t and you pass in L. How it was factored doesn't matter, but there would be no need to do so again in the function.

pec27 commented:
I think we only differ on preferences, my suggestion was just that
`multivariate_normal(mean=[0.2,0.8], cholesky=[[2,0],[1,2]], size=20)`
is more grokable than
`multivariate_normal(mean=[0.2,0.8], factored=[[2,0],[1,2]], size=20)`
even if the lower triangle didn’t have to come from the Cholesky algorithm.

bashtage commented:
I think factored would be a bool and cov would always take the value.

On Thu, Sep 14, 2017, 02:10 Peter Creasey <notifications@github.com> wrote:

> I think we only differ on preferences, my suggestion was just that
> multivariate_normal(mean=[0.2,0.8], cholesky=[[2,0],[1,2]], size=20)
> is more grokable than
> multivariate_normal(mean=[0.2,0.8], factored=[[2,0],[1,2]], size=20)
> even if the lower triangle didn’t have to come from the Cholesky algorithm.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/9561#issuecomment-329339994>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFU5RRIGFA_u80_BreCbJm9eyw9LZTxbks5siH0NgaJpZM4O3TxH>
> .
>

eric-wieser commented:
The matrix doesn't even need to be lower triangular though, does it?

Is there a multivariate analog of "standard deviation" that would be the right word here?

bashtage commented:
No. It can be full, and if full it is psd (incl symmetric).
On Thu, 14 Sep 2017 at 07:12, Eric Wieser <notifications@github.com> wrote:

> The matrix doesn't even need to be lower triangular though, does it?
>
> Is there a multivariate analog of "standard deviation" that would be the
> right word here?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/9561#issuecomment-329383694>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFU5RbKFDhDQOhS017Y5da7to5gvMY6Wks5siMPCgaJpZM4O3TxH>
> .
>

bashtage commented:
Any thoughts on value of pursuing option 3 instead of option 2?

Some more testing shows that the slowdown for the wrapped multivariate normal is almost exclusively due to the extra malloc required when first generating the normals and then calling `dot`.

charris commented:
@eric-wieser The current svd method uses full matrices, so would a method based on eigh, which I think would have been a better choice than svd if handling near singular matrices was wanted. Cholesky has the advantage of speed, but for near singular matrices it can be problematic due to roundoff errors, about the only thing one can do is raise an error.

charris commented:
> due to the extra malloc required when first generating the normals and then calling dot.

Do you mean the extra work needed to allocate the output array?

BTW, needs a rebase.

charris commented:
Hmm, it is possible to multiply in-place with BLAS (*TRMM) if one of the factors is upper/lower triangular. So there could be a cholesky fast path if we add cholesky as an optional method to `multivariate_random_normal`. Might need to improve NumPy BLAS support also.

bashtage commented:
> Do you mean the extra work needed to allocate the output array?

There is a malloc for the standard normals (unavoidable) but then since the matrix multiplication isn't inplace there is another.  At least this was what seemed to be the most expensive part (including the overhead of calling the mat multiplication)

bashtage commented:
Any thoughts on options 1 vs 2 vs 3 (really 2 vs 3 it looks like)?

charris commented:
I would choose 3) as the starting point using `multivariate_normal`. My reasoning would be

* A scalar fast path could be added together with a scalar complex
* `multivariate_normal` can be improved, with benefit to both.
* If it is worth adding broadcasting to the complex version, than `multvariate_normal` would also benefit.

The main thing that makes future modification tricky is keeping backwards compatibility of the generated sequences. Two things I'd like to add to `multivariate_normal` is a Cholesky option and a factored option. That would make factorization much faster and the multiplication could be in-place.

Some of this depends on how important broadcasting is. I have no idea how complex normals are typically used, as I've never known anyone who used them. Do you have some feedback in that regard?

charris commented:
Also, in the not to distant future, we should have more complete support for matrix multiplication of stacks with builtin broadcasting, so that should make adding those features much easier.

shoyer commented:
TensorFlow has a number of variations of multivariate normal distributions (for machine learning use-cases). It might be worth taking a look for API design inspiration. See https://www.tensorflow.org/api_docs/python/tf/contrib/distributions starting at `MultivariateNormalDiag`.

pec27 commented:
Various applications in Gaussian random fields use complex white noise (e.g. Fourier components on a periodic lattice) but these tend to be the fairly trivial cases of mean-zero complex with uncorrelated real and imaginary parts. I would be interested to know of other use cases.

bashtage committed:
CLN: Remove higher complexit implementation

Remove implementations with higher complexity in facor of simple wrapper

bashtage committed:
REF: Refactor complex notmal

Add factored to multivariate normal

bashtage commented:
Closing in favor of the future generator.
